lib/seahorn/BvOpsem2WideMemMgr.cc

```diff
@@ -0,0 +1,543 @@
+      m_ptr_sort = sort::structTy(std::move(ptr_sort), std::move(size_sort));
+    }
+
+    PtrSortTyImpl(const RawPtrSortTy &ptr_sort, const Expr &size) {
+      m_ptr_sort = sort::structTy(ptr_sort, size);
+    }
+
+    Expr v() const { return m_ptr_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    RawPtrSortTy getRawSort() { return m_ptr_sort->arg(0); }
+  };
+
+  struct MemSortTyImpl {
+    Expr m_mem_sort;
+
+    MemSortTyImpl(RawMemSortTy &&mem_sort, Expr &&size_sort) {
+      m_mem_sort = sort::structTy(std::move(mem_sort), std::move(size_sort));
+    }
+
+    MemSortTyImpl(const RawMemSortTy &mem_sort, const Expr &size_sort) {
+      m_mem_sort = sort::structTy(mem_sort, size_sort);
+    }
+
+    Expr v() const { return m_mem_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
+  using PtrTy = PtrTyImpl;
+  using PtrSortTy = PtrSortTyImpl;
+  using MemValTy = MemValTyImpl;
+  using MemSortTy = MemSortTyImpl;
+
+  /// \brief A null pointer expression (cache)
+  PtrTy m_nullPtr;
+
+  WideMemManager(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned ptrSz,
+                 unsigned wordSz, bool useLambdas = false);
+
+  ~WideMemManager() = default;
+
+  Expr bytesToSlotExpr(unsigned bytes) {
+    return m_ctx.alu().si(8 * bytes, g_slotBitWidth);
+  }
+
+  PtrSortTy ptrSort() const {
+    return PtrSortTy(OpSemMemManager::PtrSortTy(),
+                     m_ctx.alu().intTy(g_slotBitWidth));
+  }
+
+  PtrTy salloc(unsigned int bytes, uint32_t align) {
+    assert(isa<AllocaInst>(m_ctx.getCurrentInst()));
+    align = std::max(align, m_alignment);
+    auto region = m_allocator->salloc(bytes, align);
+    assert(region.first - region.second > 0);
+    return PtrTy(mkStackPtr(region.second).getRaw(),
+                 bytesToSlotExpr(region.first - region.second));
+  }
+
+  PtrTy salloc(Expr elmts, unsigned int typeSz, uint32_t align) {
+    align = std::max(align, m_alignment);
+
+    // -- compute number of bytes needed
+    Expr bytes = elmts;
+    if (typeSz > 1) {
+      // TODO: factor out multiplication and number creation
+      bytes = m_ctx.alu().doMul(bytes, m_ctx.alu().si(typeSz, ptrSzInBits()),
+                                ptrSzInBits());
+    }
+
+    // allocate
+    auto region = m_allocator->salloc(bytes, align);
+
+    // -- if allocation failed, return some pointer
+    if (m_allocator->isBadAddrInterval(region)) {
+      LOG("opsem", WARN << "imprecise handling of dynamically "
+                        << "sized stack allocation of " << *elmts
+                        << " elements of size" << typeSz << " bytes\n";);
+      return freshPtr();
+    }
+
+    // -- have a good region, return pointer to it
+    return PtrTy(
+        mkStackPtr(region.second).getRaw(),
+        m_ctx.alu().si(8 * (region.first - region.second), g_slotBitWidth));
```

Comment:
why `8 * ...`. I was expecting size to be in bytes.

Reply:
good catch!